### ISSUE TITLE
ccache -c increases reported cache size when using hard link mode

### ISSUE THREAD

**Issue by akrieger** [ccache/ccache#1320]
### How to reproduce ###

1. Do a build with ccache.
2. Use ccache -M to lower the cache size.
3. Use ccache -c to 'cleanup' the cache (even if its still undersized).

### Actual behavior ###

Cache size & file counts seemed to increase per `ccache --show-stats --verbose`.
Cache size exceeds configured maximum.

### Expected behavior ###

Cache does not increase in size when cleanup is invoked.
Cache stays under configured size limit.

### Environment ###

Ubuntu 20.04, ccache 4.8
OSX 12.6.7, ccache 4.8.2

ubuntu example, empty/small cache:
```
  ccache --show-stats --verbose
  ccache -M 2G
  ccache -c
  ccache --show-stats --verbose
  shell: /usr/bin/bash -e {0}
  env:
    ZSTD_CLEVEL: 17
    CMAKE: 0
    COMPILER: clang++-10
    MXE_TARGET: 
    WINE: 
    OS: ubuntu-20.04
    TILES: 0
    SOUND: 0
    LOCALIZE: 1
    MODS: 
    SANITIZE: 
    TEST_STAGE: 1
    CXXFLAGS: 
    LDFLAGS: 
    EXTRA_TEST_OPTS: --error-format=github-action
    NUM_TEST_JOBS: 
    NATIVE: linux64
    PCH: 1
    GOLD: 
    LTO: 
    LIBBACKTRACE: 
    RELEASE: 1
    ARCHIVE_SUCCESS: basic-build
    CCACHE_LIMIT: 2G
    CCACHE_FILECLONE: true
    CCACHE_HARDLINK: true
    CCACHE_NOCOMPRESS: true
    SKIP: false
    SKIP_TESTS: false
Cache directory:      /home/runner/.cache/ccache
Config file:          /home/runner/.config/ccache/ccache.conf
System config file:   /usr/local/etc/ccache.conf
Stats updated:        Thu Aug 10 16:37:24 2023
Cacheable calls:       589 / 592 (99.49%)
  Hits:                  0 / 589 ( 0.00%)
    Direct:              0
    Preprocessed:        0
  Misses:              589 / 589 (100.0%)
Uncacheable calls:       3 / 592 ( 0.51%)
  Called for linking:    2 /   3 (66.67%)
  No input file:         1 /   3 (33.33%)
Successful lookups:
  Direct:                0 / 589 ( 0.00%)
  Preprocessed:          0 / 589 ( 0.00%)
Local storage:
  Cache size (GB):     0.3 / 5.0 ( 6.47%)
  Files:              1767
  Hits:                  0 / 589 ( 0.00%)
  Misses:              589 / 589 (100.0%)
  Reads:              1176
  Writes:             1178
Set cache size limit to 2.0 GB
Cache directory:      /home/runner/.cache/ccache
Config file:          /home/runner/.config/ccache/ccache.conf
System config file:   /usr/local/etc/ccache.conf
Stats updated:        Thu Aug 10 16:37:34 2023
Cacheable calls:       589 / 592 (99.49%)
  Hits:                  0 / 589 ( 0.00%)
    Direct:              0
    Preprocessed:        0
  Misses:              589 / 589 (100.0%)
Uncacheable calls:       3 / 592 ( 0.51%)
  Called for linking:    2 /   3 (66.67%)
  No input file:         1 /   3 (33.33%)
Successful lookups:
  Direct:                0 / 589 ( 0.00%)
  Preprocessed:          0 / 589 ( 0.00%)
Local storage:
  Cache size (GB):     0.6 / 2.0 (30.80%)
  Files:              2945
  Hits:                  0 / 589 ( 0.00%)
  Misses:              589 / 589 (100.0%)
  Reads:              1176
  Writes:             1178
```
ubuntu example, too-large cache:
```
  ccache --show-stats --verbose
  ccache -M 2G
  ccache -c
  ccache --show-stats --verbose
  shell: /usr/bin/bash -e {0}
  env:
    ZSTD_CLEVEL: 17
    CMAKE: 0
    COMPILER: clang++-10
    MXE_TARGET: 
    WINE: 
    OS: ubuntu-20.04
    TILES: 0
    SOUND: 0
    LOCALIZE: 1
    MODS: 
    SANITIZE: 
    TEST_STAGE: 1
    CXXFLAGS: 
    LDFLAGS: 
    EXTRA_TEST_OPTS: --error-format=github-action
    NUM_TEST_JOBS: 
    NATIVE: linux64
    PCH: 1
    GOLD: 
    LTO: 
    LIBBACKTRACE: 
    RELEASE: 1
    ARCHIVE_SUCCESS: basic-build
    CCACHE_LIMIT: 2G
    CCACHE_FILECLONE: true
    CCACHE_HARDLINK: true
    CCACHE_NOCOMPRESS: true
    SKIP: false
    SKIP_TESTS: true
Cache directory:      /home/runner/.cache/ccache
Config file:          /home/runner/.config/ccache/ccache.conf
System config file:   /usr/local/etc/ccache.conf
Stats updated:        Thu Aug 10 16:36:52 2023
Cacheable calls:        589 / 592 (99.49%)
  Hits:                 233 / 589 (39.56%)
    Direct:             232 / 233 (99.57%)
    Preprocessed:         1 / 233 ( 0.43%)
  Misses:               356 / 589 (60.44%)
Uncacheable calls:        3 / 592 ( 0.51%)
  Called for linking:     2 /   3 (66.67%)
  No input file:          1 /   3 (33.33%)
Successful lookups:
  Direct:               232 / 589 (39.39%)
  Preprocessed:           1 / 357 ( 0.28%)
Local storage:
  Cache size (GB):     51.2 / 5.0 (1024.3%)
  Files:              70389
  Hits:                 235 / 591 (39.76%)
  Misses:               356 / 591 (60.24%)
  Reads:               1181
  Writes:               710
Set cache size limit to 2.0 GB
Cache directory:      /home/runner/.cache/ccache
Config file:          /home/runner/.config/ccache/ccache.conf
System config file:   /usr/local/etc/ccache.conf
Stats updated:        Thu Aug 10 16:37:00 2023
Cacheable calls:        589 / 592 (99.49%)
  Hits:                 233 / 589 (39.56%)
    Direct:             232 / 233 (99.57%)
    Preprocessed:         1 / 233 ( 0.43%)
  Misses:               356 / 589 (60.44%)
Uncacheable calls:        3 / 592 ( 0.51%)
  Called for linking:     2 /   3 (66.67%)
  No input file:          1 /   3 (33.33%)
Successful lookups:
  Direct:               232 / 589 (39.39%)
  Preprocessed:           1 / 357 ( 0.28%)
Local storage:
  Cache size (GB):     51.4 / 2.0 (2570.7%)
  Files:              70971
  Cleanups:              21
  Hits:                 235 / 591 (39.76%)
  Misses:               356 / 591 (60.24%)
  Reads:               1181
  Writes:               710
```

osx example, empty/small cache:
```
  ccache --show-stats --verbose
  ccache -M 4G
  ccache -c
  ccache --show-stats --verbose
  shell: /bin/bash -e {0}
  env:
    ZSTD_CLEVEL: 17
    CMAKE: 0
    COMPILER: clang++
    MXE_TARGET: 
    WINE: 
    OS: macos-12
    TILES: 1
    SOUND: 1
    LOCALIZE: 1
    MODS: 
    SANITIZE: 
    TEST_STAGE: 
    CXXFLAGS: 
    LDFLAGS: 
    EXTRA_TEST_OPTS: --error-format=github-action
    NUM_TEST_JOBS: 
    NATIVE: osx
    PCH: 0
    GOLD: 
    LTO: 
    LIBBACKTRACE: 
    RELEASE: 1
    ARCHIVE_SUCCESS: 
    CCACHE_LIMIT: 4G
    CCACHE_FILECLONE: true
    CCACHE_HARDLINK: true
    CCACHE_NOCOMPRESS: true
    SKIP: false
    SKIP_TESTS: false
Cache directory:            /Users/runner/Library/Caches/ccache
Config file:                /Users/runner/Library/Preferences/ccache/ccache.conf
System config file:         /usr/local/Cellar/ccache/4.8.2/etc/ccache.conf
Stats updated:              Thu Aug 10 18:21:54 2023
Cacheable calls:             587 / 590 (99.49%)
  Hits:                        0 / 587 ( 0.00%)
    Direct:                    0
    Preprocessed:              0
  Misses:                    587 / 587 (100.0%)
Uncacheable calls:             3 / 590 ( 0.51%)
  Called for linking:          2 /   3 (66.67%)
  Called for preprocessing:    1 /   3 (33.33%)
Successful lookups:
  Direct:                      0 / 587 ( 0.00%)
  Preprocessed:                0 / 587 ( 0.00%)
Local storage:
  Cache size (GB):           0.4 / 5.0 ( 7.51%)
  Files:                    1761
  Hits:                        0 / 587 ( 0.00%)
  Misses:                    587 / 587 (100.0%)
  Reads:                    1174
  Writes:                   1174
Set cache size limit to 4.0 GB
Cache directory:            /Users/runner/Library/Caches/ccache
Config file:                /Users/runner/Library/Preferences/ccache/ccache.conf
System config file:         /usr/local/Cellar/ccache/4.8.2/etc/ccache.conf
Stats updated:              Thu Aug 10 18:22:03 2023
Cacheable calls:             587 / 590 (99.49%)
  Hits:                        0 / 587 ( 0.00%)
    Direct:                    0
    Preprocessed:              0
  Misses:                    587 / 587 (100.0%)
Uncacheable calls:             3 / 590 ( 0.51%)
  Called for linking:          2 /   3 (66.67%)
  Called for preprocessing:    1 /   3 (33.33%)
Successful lookups:
  Direct:                      0 / 587 ( 0.00%)
  Preprocessed:                0 / 587 ( 0.00%)
Local storage:
  Cache size (GB):           0.7 / 4.0 (17.24%)
  Files:                    2348
  Hits:                        0 / 587 ( 0.00%)
  Misses:                    587 / 587 (100.0%)
  Reads:                    1174
  Writes:                   1174
```

osx example, too large cache:
```
  ccache --show-stats --verbose
  ccache -M 4G
  ccache -c
  ccache --show-stats --verbose
  shell: /bin/bash -e {0}
  env:
    ZSTD_CLEVEL: 17
    CMAKE: 0
    COMPILER: clang++
    MXE_TARGET: 
    WINE: 
    OS: macos-12
    TILES: 1
    SOUND: 1
    LOCALIZE: 1
    MODS: 
    SANITIZE: 
    TEST_STAGE: 
    CXXFLAGS: 
    LDFLAGS: 
    EXTRA_TEST_OPTS: --error-format=github-action
    NUM_TEST_JOBS: 
    NATIVE: osx
    PCH: 0
    GOLD: 
    LTO: 
    LIBBACKTRACE: 
    RELEASE: 1
    ARCHIVE_SUCCESS: 
    CCACHE_LIMIT: 4G
    CCACHE_FILECLONE: true
    CCACHE_HARDLINK: true
    CCACHE_NOCOMPRESS: true
    SKIP: false
    SKIP_TESTS: true
Cache directory:            /Users/runner/Library/Caches/ccache
Config file:                /Users/runner/Library/Preferences/ccache/ccache.conf
System config file:         /usr/local/Cellar/ccache/4.8.2/etc/ccache.conf
Stats updated:              Thu Aug 10 17:03:37 2023
Cacheable calls:              587 / 590 (99.49%)
  Hits:                       232 / 587 (39.52%)
    Direct:                   232 / 232 (100.0%)
    Preprocessed:               0 / 232 ( 0.00%)
  Misses:                     355 / 587 (60.48%)
Uncacheable calls:              3 / 590 ( 0.51%)
  Called for linking:           2 /   3 (66.67%)
  Called for preprocessing:     1 /   3 (33.33%)
Successful lookups:
  Direct:                     232 / 587 (39.52%)
  Preprocessed:                 0 / 355 ( 0.00%)
Local storage:
  Cache size (GB):           10.6 / 5.0 (211.2%)
  Files:                    21517
  Hits:                       232 / 587 (39.52%)
  Misses:                     355 / 587 (60.48%)
  Reads:                     1174
  Writes:                     710
Set cache size limit to 4.0 GB
Cache directory:            /Users/runner/Library/Caches/ccache
Config file:                /Users/runner/Library/Preferences/ccache/ccache.conf
System config file:         /usr/local/Cellar/ccache/4.8.2/etc/ccache.conf
Stats updated:              Thu Aug 10 17:03:46 2023
Cacheable calls:              587 / 590 (99.49%)
  Hits:                       232 / 587 (39.52%)
    Direct:                   232 / 232 (100.0%)
    Preprocessed:               0 / 232 ( 0.00%)
  Misses:                     355 / 587 (60.48%)
Uncacheable calls:              3 / 590 ( 0.51%)
  Called for linking:           2 /   3 (66.67%)
  Called for preprocessing:     1 /   3 (33.33%)
Successful lookups:
  Direct:                     232 / 587 (39.52%)
  Preprocessed:                 0 / 355 ( 0.00%)
Local storage:
  Cache size (GB):           10.8 / 4.0 (270.5%)
  Files:                    21872
  Hits:                       232 / 587 (39.52%)
  Misses:                     355 / 587 (60.48%)
  Reads:                     1174
  Writes:                     710
```

**Comment by jrosdahl** [comment]
Thanks for the bug report.

Note that the only thing `ccache -M` does is to write a new `max_size` value to the configuration file. `ccache -c` does two things: Recalculates file and size counters (stored in `<ccache_dir>/X/Y/stats` files) and then deletes cache entries until the cache is within the size limit.

In the _ubuntu example, empty/small cache_ case, it seems like `ccache -c` found more files in the cache than what was previously noted in the `stats` files (_Files_ went from 1767 to 2945). That is unexpected, but you should see the same effect without a preceding `ccache -M` call. (To clarify: what's unexpected is that there are more files than noted in the `stats` files, not that `ccache -c` finds those files.)

In the _ubuntu example, too-large cache_ case, more files were found just like described above, but I can't explain why `ccache -c` did not trim the cache down to 2 GiB. Would it be possible for me to get a copy of that cache directory for inspection?

I cannot reproduce the behavior you're seeing using the reproduction steps. Can you (outside your CI pipeline)?

My gut feeling is that the "found more files" part of the problem has to do with some interaction with `actions/cache` and how it saves or restores things. For instance, could it be that you're building some things in the _prepare_ step using ccache, thus adding entries to the ccache cache? If so, then I'm guessing that the _ccache cache files_ step will overwrite the ccache `stats` files, which would explain why the files/size counters are incorrect and why `ccache -c` will find more files than before.

**Comment by akrieger** [comment]
Here is the log archive for the 'small ubuntu' case (actual instance is https://github.com/CleverRaven/Cataclysm-DDA/actions/runs/5823222887/job/15789649067, but I expect it'll expire soon). That build instance is from a PR where I change the cachekey, inducing a miss and dropping the entire preexisting cache. Prior to build, all  we do is a checkout and some apt install calls.

[logs_855474.zip](https://github.com/ccache/ccache/files/12362558/logs_855474.zip)

**Comment by jrosdahl** [comment]
I just noticed the detail I needed to be able to reproduce the problem: it's because you're using hard links, and the new cleanup algorithm makes an assumption that isn't true about how bookkeeping is done when using hard links. Essentially, it's a bug in the hard link mode bookkeeping that wasn't exposed until now but happened to work by luck with the old cleanup algorithm. I'll have to think a bit about how to fix it.
